src/plugin.js

```diff
@@ -0,0 +1,143 @@
+   */
+  async loadModules() {
+    // load a list of modules from file path or node_modules
+    const loadPlugin = (await lpModule).loadPlugin;
+    const resolvePlugin = (await lpModule).resolvePlugin;
+    if (this.names !== undefined && this.names.length !== 0) {
+      console.log(`Found ${this.names.length} packages to load.`);
+      this.names.forEach(async (name) => {
+        try {
+          // Only plugins from the @finos scope are supported
+          const resolved = await resolvePlugin(name, { prefix: '@finos' });
+          const plugin = await loadPlugin(resolved, { prefix: '@finos' });
+          this.pluginModules.concat(plugin);
+        } catch (err) {
+          console.error(`Unable to load plugin ${name}: ${err}`);
+        }
+      });
+    }
+    if (this.paths !== undefined && this.paths.length !== 0) {
+      console.log(`Found ${this.paths.length} local plugins to load.`);
+      this.paths.forEach(async (p) => {
+        try {
+          const resolved = await resolvePlugin(path.join(process.cwd(), p));
+          const plugin = await loadPlugin(resolved);
+          this.pluginModules.concat(plugin);
+        } catch (err) {
+          console.error(`Unable to load plugin ${p}: ${err}`);
+        }
+      });
+    }
+    this.modulesLoaded = true;
+  }
+
+  loadPlugins() {
+    // This function runs before loadModules finishes. No idea how to resolve
+    // since load-plugin package only supports async.
+    const plugins = [];
+    for (pluginModule in this.pluginModules) {
+      // iterate over the module.exports keys
+      for (key in Object.keys(pluginModule)) {
+        // only extract instances of the GenericPlugin base class
+        // add it to the list of loaded plugins
+        if (
+          Object.prototype.hasOwnProperty.call(pluginModule, key) &&
+          p[key] instanceof ProxyPlugin
+        ) {
+          plugins.concat(p[key]);
+        }
+      }
+    }
+    this.plugins = plugins;
+    this.loaded = true;
+  }
+}
+
+/**
+ * Parent class for all Git Proxy plugins. New plugin types must inherit from
+ * ProxyPlugin to be loaded by PluginManager.
+ */
+class ProxyPlugin {}
+
+/**
+ * A plugin which executes a function when receiving a proxy request.
+ */
+class GenericPlugin extends ProxyPlugin {
+  /**
+   * Setup a new instance of GenericPlugin with a custom exec function.
+   * @param {function} execFunc A function that executes when a push is proxied.
+   */
+  constructor(execFunc) {
+    super();
+    this.pluginType = 'generic';
+    this.execFunc = execFunc;
+  }
+
+  /**
+   *
+   * @param {*} request
+   * @param {*} action
+   */
+  execute(request, action) {
+    this.execFunc(request, action);
+  }
+}
+
+const setupPluginManager = async () => {
+  // Auto-register plugins that are part of git-proxy core
+  let packagePlugins = [];
+  let localPlugins = ['./packages/git-proxy-notify-hello/index.js'];
+  // How should additional plugins be registered? Environment variable?
```

Comment:
Isn't it what `(process.env.GITPROXY_LOAD_PLUGINS)` is for?

Reply:
That's the thought for how _user-specified_ plugins can be set. As part of core git-proxy, it's not unreasonable to dog-food and load internal plugins from a static list. In this case, I'm using a simple "hello world" example but thinking this can be done for many existing components within the proxy itself that can be modularized.

Anything specified via the environment variable will be additive (in addition to "core" or "bundled" plugins). Open to other methods though.

Reply:
I guess I'm posing the question in that comment on if there is a better wait to support user plugins without having to scan an entire working directory or the local `node_modules/`. Worth iterating on in the future but for now, an environment variable is fine.